db/redis/db.go

```diff
@@ -0,0 +1,366 @@
+package redis
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"strings"
+	"time"
+
+	goredis "github.com/go-redis/redis"
+	"github.com/magiconair/properties"
+	"github.com/pingcap/go-ycsb/pkg/prop"
+	"github.com/pingcap/go-ycsb/pkg/ycsb"
+)
+
+func init() {
+	ycsb.RegisterDBCreator("redis", redisCreator{})
+}
+
+type redis struct {
+	singleCli  *goredis.Client
+	clusterCli *goredis.ClusterClient
+	mode       string
+}
+
+func (r *redis) Close() error {
+	var err error
+	switch r.mode {
+	case "cluster":
+		err = r.clusterCli.Close()
+	case "single":
+		fallthrough
+	default:
+		err = r.singleCli.Close()
+	}
+	return err
+}
+
+func (r *redis) InitThread(ctx context.Context, _ int, _ int) context.Context {
+	return ctx
+}
+
+func (r *redis) CleanupThread(_ context.Context) {
+}
+
+func (r *redis) Read(ctx context.Context, table string, key string, fields []string) (map[string][]byte, error) {
+	var res string
+	var err error
+
+	data := map[string][]byte{}
+
+	switch r.mode {
+	case "cluster":
+		res, err = r.clusterCli.Get(table + "/" + key).Result()
+		if err != nil {
+			return nil, err
+		}
+	case "single":
+		fallthrough
+	default:
+		res, err = r.singleCli.Get(table + "/" + key).Result()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	err = json.Unmarshal([]byte(res), &data)
+	if err != nil {
+		return nil, err
+	}
+
+	return data, err
+}
+
+var cursor uint64
+
+func (r *redis) Scan(ctx context.Context, table string, startKey string, count int, fields []string) ([]map[string][]byte, error) {
+	data := []map[string][]byte{}
+	var keys []string
+	var err error
+
+	switch r.mode {
+	case "cluster":
+		keys, cursor, err = r.clusterCli.Scan(cursor, table+"/*", int64(count)).Result()
+		if err != nil {
+			return nil, err
+		}
+
+		for _, k := range keys {
+			res, err := r.clusterCli.Get(k).Result()
+			if err != nil {
+				return nil, err
+			}
+
+			tmp := map[string][]byte{}
+			err = json.Unmarshal([]byte(res), &tmp)
+			if err != nil {
+				return nil, err
+			}
+
+			data = append(data, tmp)
+		}
+	case "single":
+		fallthrough
+	default:
+		keys, cursor, err = r.singleCli.Scan(cursor, table+"/*", int64(count)).Result()
+		if err != nil {
+			return nil, err
+		}
+
+		for _, k := range keys {
+			res, err := r.singleCli.Get(k).Result()
+			if err != nil {
+				return nil, err
+			}
+
+			tmp := map[string][]byte{}
+			err = json.Unmarshal([]byte(res), &tmp)
+			if err != nil {
+				return nil, err
+			}
+
+			data = append(data, tmp)
+		}
+	}
+
+	return data, nil
+}
+
+func (r *redis) Update(ctx context.Context, table string, key string, values map[string][]byte) error {
+	data, err := json.Marshal(values)
```

Comment:
you can't set this data directly, the update values only contain the fields need to be updated later, you need to get the old values, update them.

Reply:
Ah, I didn't know that. I was assuming you give me the whole new value.

Reply:
Fixed in https://github.com/pingcap/go-ycsb/pull/89/commits/85a0affd7735da239a46d1b4d17428fa6f7ce7cc